### PR TITLE
[FLINK-28419][table] Add runtime provider interface for full caching lookup + implement Periodic and Timed cache reload triggers

### PR DESCRIPTION
**Pull request by SmirAlex**
## What is the purpose of the change

This pull request introduces new interfaces that allows to use Full cache as `LookupFunctionProvider`. Interface list:

- `FullCachingLookupProvider`
- `CacheReloadTrigger`

CacheReloadTrigger provides customizable way when to trigger cache reload operation. There are 2 built-in triggers:

- `PeriodicCacheReloadTrigger` - cache is reloaded with fixed intervals without initial delay
- `TimedCacheReloadTrigger` - cache is reloaded at specified time with fixed intervals multiples of one day

Also there are introduced common options for connectors that will support Full caching lookup join.

## Brief change log

- Add interfaces FullCachingLookupProvider and CacheReloadTrigger
- Implement reload triggers PeriodicCacheReloadTrigger and TimedCacheReloadTrigger
- Add common SQL options for Full caching lookup join

## Verifying this change

This change added unit tests for the newly added `PeriodicCacheReloadTrigger` and `TimedCacheReloadTrigger`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (FLIP docs + JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/CacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/CacheReloadTrigger.java:37]: Add @PublicEvolving here

**Review by PatrickRen — COMMENTED** [review]
@SmirAlex Thanks for the contribution! The PR overall looks good to me. Just some minor comments.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTriggerTest.java
Comment (SmirAlex) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTriggerTest.java]: Yea, I just thought this little one would be ok, because basically we don't need any implementation of this Runnable, except counting calls of one method. Anyway I fixed that, no problem.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/CacheReloadTrigger.java
Comment (SmirAlex) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/CacheReloadTrigger.java:38]: Fixed

**Comment by SmirAlex** [comment]
@PatrickRen  Thanks for the review! I've made fixes according to your comments. Please look at the latest commit, when you will have time.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:42]: The serial version UID should start from 1 for new classes

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:103]: What about wrapping with a `if(scheduledExecutor != null)`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java:46]: The serial version UID should start from 1 for new classes

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java:110]: What about wrapping with a if(scheduledExecutor != null)?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/LookupOptions.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/LookupOptions.java:48]: I'm not sure if it is a good idea to have a default value with 0 here. If users intend to reload the cache periodically, the interval should be set in the configuration to fully express the reloading strategy. Maybe we could just leave it without default value here and have a check in the `PeriodicCacheReloadTrigger`. WDYT?

**Review by PatrickRen — COMMENTED** [review]
Thanks for the update @SmirAlex ! I left some comments. Also please rebase the latest master as the `LookupOptions` has been merged to master.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/LookupOptions.java
Comment (SmirAlex) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/LookupOptions.java]: Yes, I agree. If partial cache can't be built without specifying options for it, the same should be for full cache.

**Comment by SmirAlex** [comment]
@PatrickRen I resolved your comments and rebased. Additionally I changed error messages and config descriptions, so they now correspond to the ones for partial caching. You can have a look. Do I need to run CI again before merge?

**Comment by PatrickRen** [comment]
@SmirAlex Yeah we have to wait for a green CI run. It should be running now so let's wait for a while.

**Comment by SmirAlex** [comment]
@flinkbot run azure

**Comment by SmirAlex** [comment]
@PatrickRen There were multiple failed pipelines, but they were because of unrelated issues. Latest pipeline was successful - https://dev.azure.com/asmirnov1/Flink/_build/results?buildId=17&view=results

**Review by PatrickRen — APPROVED** [review]
Thanks for the contribution @SmirAlex ! LGTM

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/FullCachingLookupProvider.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/FullCachingLookupProvider.java:34]: Missing JavaDoc for this public API method

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java
Comment (PatrickRen) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:39]: I think this JavaDoc should be updated as the argument check in the constructor has disabled the possibility of using zero as the reload interval.

**Review by PatrickRen — COMMENTED** [review]
Just two minor comments about JavaDoc.

**Comment by SmirAlex** [comment]
@PatrickRen thanks for the review! I've fixed last comments about javadoc.

**Comment by PatrickRen** [comment]
Merged on master 5405239dec0884dff746129c73955c90f455c465

**Comment by xiaolan-bit** [comment]
I would like to ask, as I have a scenario where Flink writes into Hudi. Due to the complexity of the SQL process, I want to turn the whole process into an SQL file. However, due to the latency of the insert operation, I want to delay the operation through Flink SQL statements. Is there such a statement that can implement the delay operation, for example, execute the next SQL statement only after 5 seconds?

**Pull request closed without merging**
